packages/@aws-cdk/aws-s3/README.md

```diff
@@ -312,3 +312,20 @@ const bucket = new Bucket(this, 'MyRedirectedBucket', {
 To put files into a bucket as part of a deployment (for example, to host a
 website), see the `@aws-cdk/aws-s3-deployment` package, which provides a
 resource that can do just that.
+
+### Bucket deletion
+
+By default, when a bucket is removed from a stack (or the stack is deleted),
+the S3 bucket will be removed according to its removal policy (which by
+default, will simply orphan the bucket and leave it in your AWS account).
+However, even if the removal policy is set to DESTROY, the bucket will get
+skipped by default if it still has objects inside.
+
+To override this, enable the `autoDeleteObjects` property.
+
+```ts
+const bucket = new Bucket(this, 'MyTempFileBucket', {
+  removalPolicy: RemovalPolicy.DESTROY,
+  autoDeleteObjects: true,
```

Comment:
Honestly I think there should be no need for an extra flag, given the behavior of removalPolicy: Destroy is pretty obvious.

Reply:
An alternative would be to enrich the removalPolicy with a third option: `DESTROY_AND_DELETE_CONTENTS`.

I've seen such "construct specific" values available for removalPolicy in other CDK constructs, like SNAPSHOT for the RDS construct.

Reply:
> If this is indeed the behavior, we cannot change the meaning of removalPolicy: DESTROY because it might result in an unexpected object deletion as far the user is concerned.

Yep - I think you hit the nail on the head here. Right now users do not expect that buckets set to "DESTROY" will actually destroy any data, since the deletion operation will only succeed if it is empty.

> I am more comfortable with piggy backing on the DESTROY to infer object deletion, since this scenario is essentially broken anyway.  

I'm still a little hesitant just because it might surprise some users -- and even if we implement this feature as a default for all buckets with `RemovalPolicy.DESTROY`, I imagine some users would still want the ability to opt out of the additional-resource-per-stack overhead created by this feature, which might be merit for some kind of flag of its own. It's an open question as to whether this should be an opt-in or opt-out feature. That said I'm totally open to piggy backing on `DESTROY` if that makes more sense to most users and the cdk team! Thoughts?

> An alternative would be to enrich the removalPolicy with a third option: `DESTROY_AND_DELETE_CONTENTS`.
> 
> I've seen such "construct specific" values available for removalPolicy in other CDK constructs, like SNAPSHOT for the RDS construct for example.

This is a nice solution idea! But I'm not sure if it would work since SNAPSHOT is specifically a deletion policy provided by CloudFormation (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html#aws-attribute-deletionpolicy-options) - where as we would be adding something specific to the cdk library. We could add a fourth unofficial enum option to the `RemovalPolicy` class as you suggest, but it might clutter the API and require adding extra logic for handling this extra case in other parts of the library.

I did see there is a related interface named "RemovalPolicyOptions", but I think its purpose is just to modify the underlying synthesized CfnResources - whereas we are trying to add a more high-level behavior which is specific to buckets. So I don't think it's the right tool for our purposes, but I could be wrong.

> So, question is - what exactly did you mean by "the bucket will be skipped"?  

Ah, that's a good question. 😅 I'm not entirely sure since I wrote this a few months ago, but I think in the past (or maybe under different conditions) the bucket would switch to a "DELETE_SKIPPED" state instead of "DELETE_FAILED". I did find a reference to this behavior here (https://github.com/aws/aws-cdk/issues/2601#issuecomment-498850970). But today I did the same test as you, and it looks like this is no longer the case, so I think my doc needs to be updated. Thanks for catching this :)

Reply:
Ok, I feel this is a little too fragile/critical to tinker with. Lets just keep the current implementation of using an opt-in flag 👍